tensorboard/components/tensor_widget/tensor-widget-impl.ts

```diff
@@ -146,12 +178,388 @@ export class TensorWidgetImpl implements TensorWidget {
+        if (i < (values as number[]).length) {
+          valueDiv.textContent = numericValueToString(
+            values[i],
+            isIntegerDType(this.tensorView.spec.dtype)
+          );
+        } else {
+          valueDiv.textContent = '';
+        }
+      }
+    } else if (this.rank >= 2) {
+      for (let i = 0; i < numRows; ++i) {
+        for (let j = 0; j < numCols; ++j) {
+          const valueDiv = this.valueDivs[i][j];
+          if (
+            i < (values as number[][]).length &&
+            j < (values[i] as number[]).length
+          ) {
+            valueDiv.textContent = numericValueToString(
+              values[i][j],
+              isIntegerDType(this.tensorView.spec.dtype)
+            );
+          } else {
+            valueDiv.textContent = '';
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Fill in the content of the top/left rulers and main value divs
+   * based on the current slicing spec.
+   */
+  private async renderRulersAndValueDivs() {
+    this.renderTopRuler();
+    this.renderLeftRuler();
+    await this.renderValueDivs();
+  }
+
+  /**
+   * Scroll horizontally to a specified column index.
+   *
+   * This is a no-op for scalar (0D) and 1D tensors.
+   * If the column index is out of bound under the current slicing spec,
+   * an error will be thrown.
+   *
+   * @param index
+   */
   async scrollHorizontally(index: number) {
-    throw new Error('scrollHorizontally() is not implemented yet.');
+    if (this.rank <= 1) {
+      // Cannot scroll the display of a scalar or 1D tensor.
+      return;
+    }
+    const indexUpperBound = this.tensorView.spec.shape[
+      this.slicingSpec.viewingDims[1]
+    ];
+    if (index < 0 || index >= indexUpperBound) {
+      throw new Error(
+        `Index out of bound: ${index} is outside [0, ${indexUpperBound}})`
+      );
+    }
+
+    this.slicingSpec.horizontalRange[0] = index;
+    this.slicingSpec.horizontalRange[1] = index + this.topRulerTicks.length;
+    const maxCol = this.tensorView.spec.shape[this.slicingSpec.viewingDims[1]];
+    if (this.slicingSpec.horizontalRange[1] > maxCol) {
+      this.slicingSpec.horizontalRange[1] = maxCol;
+    }
+
+    await this.renderRulersAndValueDivs();
   }
 
+  /**
+   * Scroll vertically to a specified row index.
+   *
+   * This is a no-op for scalar (0D) tensors.
+   * If the row index is out of bound under the current slicing spec,
+   * an error will be thrown.
+   *
+   * @param index
+   */
   async scrollVertically(index: number) {
-    throw new Error('scrollVertically() is not implemented yet.');
+    if (this.rank === 0) {
+      // Cannot scroll the display of a scalar.
+      return;
+    }
+    const indexUpperBound = this.tensorView.spec.shape[
+      this.slicingSpec.viewingDims[0]
+    ];
+    if (index < 0 || index >= indexUpperBound) {
+      throw new Error(
+        `Index out of bound: ${index} is outside [0, ${indexUpperBound}})`
```

Comment:
Syntax error on this line?  Seems there should be a square bracket instead of a curly brace

Reply:
This is a typo, but not a TypeScript syntax error, which is why it wasn't caught. Fixed now. Thanks for catching it.